### PR TITLE
[FEATURE] Implement Experimental SparkDatasource with CSVDataAsset

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
This change introduces Experimental `SparkDatasource` with `CSVDataAsset` (for now, `CSVAsset`).  Here are some notes:
* We need to improve type/name/extension management for `DataAsset` registration so as to make user-accessible classes/methods consistent.  (There is a "TODO" note for it.)
* We need to figure out a formal way to handle `reader_options` for `Spark` in particular and different options for various `DataAsset` implementations in particular; the current "work-around" hard-codes the needed ones into the `CSVSparkAsset` implementation.  (There is a "TODO" note for it.)
* Integration tests have been refactored to use a utility module and common `DataContext` and `DataAsset` loading methods from stored fixtures were moved into `conftest.py` and turned into `PyTest` fixtures, where appropriate.

### Additional changes
* Rename `logger` to `LOGGER` per `Python` style guidelines.
* Delete unused `MetricFunctionTypes` `Enum` constants (only `VALUE` is used).

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-579/GREAT-1603
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: tests/experimental/datasources/test_spark_datasource.py
Comment (cdkini) [inline tests/experimental/datasources/test_spark_datasource.py]: We should `pytest.skip` if we don't have Spark in our `test_backends`

File: tests/experimental/datasources/test_spark_datasource.py
Comment (cdkini) [inline tests/experimental/datasources/test_spark_datasource.py:40]: If using a true Spark connection, I don't believe these are unit tests?

**Review by cdkini — DISMISSED** [review]

File: great_expectations/core/util.py
Comment (cdkini) [inline great_expectations/core/util.py]: Are these comments necessary? Can we remove them?

File: great_expectations/experimental/datasources/interfaces.py
Comment (cdkini) [inline great_expectations/experimental/datasources/interfaces.py]: Is there a ticket to capture this? If not, let's do so because code comments tends to get lost in the mix

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (cdkini) [inline great_expectations/experimental/datasources/spark_datasource.py]: Nit - why do we use `LOGGER` some places and `logger` elsewhere? Can we standardize?

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (cdkini) [inline great_expectations/experimental/datasources/spark_datasource.py]: Same note as before - for any code comments, do they make more sense as Jira tickets or discussion points at arch review or in our Slack? We have many outdated comments in our codebase - I'm wary of adding more if they aren't necessary

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (cdkini) [inline great_expectations/experimental/datasources/spark_datasource.py:85]: Nit - we don't need to explicitly type here (or anywhere else where the type can be implied. `mypy` dislikes these annotations unless there is ambiguity present

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (cdkini) [inline great_expectations/experimental/datasources/spark_datasource.py:259]: Are all these `type: ignore` comments necessary? Is there a way to get around it?

File: tests/experimental/datasources/test_utils.py
Comment (cdkini) [inline tests/experimental/datasources/test_utils.py]: Do we really need a utils file? Could these not be moved to fixture or ported over to the test files where they are used?

I personally think its very difficult to know what utils exist

File: great_expectations/core/util.py
Comment (alexsherstinsky) [reply great_expectations/core/util.py]: @cdkini This was debug -- deleting (in next commit).

File: great_expectations/experimental/datasources/interfaces.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/interfaces.py]: @cdkini Good idea -- I will create the JIRA ticket.  Thanks.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py]: @cdkini I am not sure, but in ZEP the style has been capitalized "LOGGER" -- perhaps @Kilo59 and/or @billdirks could provide more background.  Thank you.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py]: @cdkini Good idea -- I will create the JIRA ticket.  Thanks.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py:68]: @cdkini Could you please suggest a better coding pattern?  Thank you.

Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py:85]: @cdkini Thank you -- I will take a look (including `Pandas`).  This code is temporary anyway.  Thanks.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py:259]: @cdkini I will check (they were in the `Pandas` version).  Thanks.

File: tests/experimental/datasources/test_utils.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_utils.py]: @cdkini I was thinking of them as utilities independent of fixtures, but I will double-check and combine if it can be done.  Thanks.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py:85]: @cdkini I checked and feel that the type hint is useful here (and in `Pandas`), because of what the list contains (i.e., `pathlib.Path` is the non-obvious part here).  Thanks.

File: tests/experimental/datasources/test_spark_datasource.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_spark_datasource.py]: @cdkini Applied (committed).  Thanks!

File: tests/experimental/datasources/test_spark_datasource.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_spark_datasource.py:40]: @cdkini I am not sure whether or not using the `pyspark` module in our local (non-distributed) tests (even in Azure) qualifies as a "true" connection.  So I am inclined to keep as is.  Thanks.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py:259]: @cdkini I checked.  Without these "ignore" comments, the following happened:

```
mypy great_expectations/experimental/datasources/spark_datasource.py
great_expectations/experimental/datasources/spark_datasource.py:225: error: Argument "base_directory" to "CSVSparkAsset" has incompatible type "Union[str, Path]"; expected "Path"  [arg-type]
great_expectations/experimental/datasources/spark_datasource.py:226: error: Argument "regex" to "CSVSparkAsset" has incompatible type "Union[str, Pattern[Any]]"; expected "Pattern[Any]"  [arg-type]
great_expectations/experimental/datasources/spark_datasource.py:227: error: Argument "order_by" to "CSVSparkAsset" has incompatible type "Union[List[BatchSorter], List[str], List[<nothing>]]"; expected "List[BatchSorter]"  [arg-type]
Found 3 errors in 1 file (checked 1 source file)
```

With these "ignores" in place, `mypy` ran cleanly.

Thanks.

File: tests/experimental/datasources/conftest.py
Comment (cdkini) [inline tests/experimental/datasources/conftest.py]: Should these be fixtures by the way?

File: tests/experimental/datasources/conftest.py
Comment (cdkini) [inline tests/experimental/datasources/conftest.py]: Since we have this return type everywhere, maybe we use a namedtuple? Not super necessary but might be a nice quality-of-life improvement

File: tests/experimental/datasources/test_spark_datasource.py
Comment (cdkini) [reply tests/experimental/datasources/test_spark_datasource.py:40]: I don't know the intricacies of this new datasource (or ZEP in general) so I'm going to defer to Gabe and Bill - I want to say these should be integration though

File: tests/experimental/datasources/test_utils.py
Comment (cdkini) [inline tests/experimental/datasources/test_utils.py]: These numbers should probably be constants and defined at the top of the file?

File: tests/experimental/datasources/test_utils.py
Comment (cdkini) [reply tests/experimental/datasources/test_utils.py]: there feels like there is quite a lot of logic here - I think all of these calls will result in integration tests. Please ensure that any uses are appropriately marked

File: tests/experimental/datasources/test_utils.py
Comment (cdkini) [inline tests/experimental/datasources/test_utils.py]: Is all of this logic necessary? This feels like so much to assert against - what is the purpose of our tests? Can we narrow our scope at all so we don't have brittle tests?

File: tests/experimental/datasources/test_utils.py
Comment (cdkini) [inline tests/experimental/datasources/test_utils.py]: On second thought, I'm going to push for a namedtuple here - this is difficult to reason about as a len 4 tuple in my opinion

File: tests/experimental/datasources/test_utils.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_utils.py]: @cdkini This is not new (came from before Spark).  However, I believe that it is temporary and will be replaced with a more fully-featured and performant implementation soon.  Thanks.

File: tests/experimental/datasources/test_utils.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_utils.py]: @cdkini Again, this is not new.  We can update this implementation in a subsequent effort -- however, it is temporary anyway and will be replaced by a fuller file_path handling implementation soon.

File: tests/experimental/datasources/test_spark_datasource.py
Comment (cdkini) [reply tests/experimental/datasources/test_spark_datasource.py:40]: The very requirement on connecting to Spark (even if that connection is not used) makes me think that unit is not applicable. Since we're taking a lot of care in crafting our unit tests, let's be confident in our approach here and be consistent

Comment (alexsherstinsky) [reply tests/experimental/datasources/test_spark_datasource.py:40]: @cdkini @Kilo59 @billdirks -- Fine with me either way; I would just want to appeal for consistency (if makes sense).  If this test is integration, then so should be the same tests for the other backends.  Thanks.

File: tests/experimental/datasources/test_utils.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_utils.py]: @cdkini This is refactoring (not new code).  I will look to advice from @Kilo59 and @billdirks in terms of which tests should be marked as `@integration`.  Thanks!

File: tests/experimental/datasources/conftest.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/conftest.py]: @cdkini This function, and the appropriately-named functions for the other backends and their multi-batch counterparts, are all used in fixtures.  However, making them -- in addition -- fixtures themselves causes problems with `pytest` which is notoriously known for not accepting fixtures as parameters to other fixtures.  (I tried, and after error messages explaining the above, reverted to just declaring them as methods to be used in fixtures.)  If you know of a way, please do let me know.  Thank you.

File: tests/experimental/datasources/conftest.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/conftest.py]: @cdkini This was written in the same style as the analogous `pandas_data()` method.

File: tests/experimental/datasources/test_utils.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_utils.py]: @cdkini We have not done that in other tests that assert on specific counts (e.g., in the `DataAssistant` tests).  Thanks.

File: tests/experimental/datasources/test_spark_datasource.py
Comment (Kilo59) [reply tests/experimental/datasources/test_spark_datasource.py:40]: I'm fine with changing it.
This is just an old `pylint` habit.
https://pylint.pycqa.org/en/latest/user_guide/messages/convention/invalid-name.html#invalid-name-c0103

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (Kilo59) [reply great_expectations/experimental/datasources/spark_datasource.py]: I'm fine with changing it but...
This is an old `pylint` habit. Pylint wants any model level constants to be `UPPER_CASE`.
>`const` - Module-level constants, any variable defined at module level that is not bound to a class object.
https://pylint.pycqa.org/en/latest/user_guide/messages/convention/invalid-name.html#invalid-name-c0103

This is also the google style recommendation, but it doesn't appear that we've been following it.
https://google.github.io/styleguide/pyguide.html#254-decision

Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py]: @Kilo59 @cdkini @billdirks -- I am happy to convert all of our `logger.` usage to `LOGGER.` in an immediately following pull request -- after the present pull request is merged with `logger.` spelled as is in this pull request's working branch right now.  Thanks.

File: great_expectations/experimental/datasources/dynamic_pandas.py
Comment (billdirks) [inline great_expectations/experimental/datasources/dynamic_pandas.py:126]: Tagging @Kilo59 so he knows this is changing.

File: great_expectations/experimental/datasources/pandas_datasource.py
Comment (billdirks) [inline great_expectations/experimental/datasources/pandas_datasource.py:279]: nit: remove 'relative' so it is not confused with 'relative path'. That is change it to something like: `The path to the directory where the csv files will be collected.`

File: great_expectations/experimental/datasources/pandas_datasource.py
Comment (billdirks) [inline great_expectations/experimental/datasources/pandas_datasource.py]: This `order_by` description doesn't match with the type, `Optional[BatchSortersDefinition]`, above.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/spark_datasource.py]: +1 remove this comment and file a jira ticket instead.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (billdirks) [inline great_expectations/experimental/datasources/spark_datasource.py:79]: The file system, regex matching code is datasource agnostic so it feels like it should be moved to some shared module that the pandas and pyspark implementations can reuse. I think I'll like that to be down now, but we could wait until we need to do edits in both places.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (billdirks) [inline great_expectations/experimental/datasources/spark_datasource.py]: Could you file a JIRA and delete this comment? Why is this required now? What does a fully-functional solution do that this doesn't? ie what are the acceptance criteria for a fix.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (billdirks) [inline great_expectations/experimental/datasources/spark_datasource.py:35]: This whole class is a lot of copy and paste from the pandas code. Could we refactor this functionality into some shared code that gets called from this class.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (billdirks) [inline great_expectations/experimental/datasources/spark_datasource.py]: This doesn't appear to be correct. It looks like it should be a list of fields that one wants to sort by with an optional prefix of `+` or `-` if one wants to specify sort direction.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (billdirks) [reply great_expectations/experimental/datasources/spark_datasource.py:259]: This has to do with pydantic run time type coercion that mypy doesn't know about.

File: tests/experimental/datasources/conftest.py
Comment (billdirks) [inline tests/experimental/datasources/conftest.py:1]: I'm curious why you broke these out of the `test_integration_datasources.py` file. These functions are about connecting to real datasources and should only be called integration tests. We should actively discourage there use elsewhere.
If we want to break up integration tests maybe we should have an `integration` directory and put integration tests there and shared utilities under that directory.

File: tests/experimental/datasources/test_spark_datasource.py
Comment (billdirks) [inline tests/experimental/datasources/test_spark_datasource.py]: Could you get rid of the `file_relative_path` dependency and change this line to:
```
return pathlib.Path(__file__) / "../../../test_sets/taxi_yellow_tripdata_samples"
```

File: tests/experimental/datasources/test_utils.py
Comment (billdirks) [reply tests/experimental/datasources/test_utils.py]: Both the code here and in the new `conftest.py` came from `test_integration_datasource.py`. All the functionality should only be used in integration tests and we should discourage its use elsewhere. If that file was getting to big, please see my comment there for a strategy for doing this.

File: tests/experimental/datasources/test_spark_datasource.py
Comment (billdirks) [reply tests/experimental/datasources/test_spark_datasource.py:40]: I don't believe any of these tests actually need the spark backend. We might instantiate one when the datasource comes up but the tests check the `Batch` (a dataclass of data describing the data in the batch) which might involve listing the csv files in a directory. We could mock the spark execution engine to more strongly separate this from using actual spark. For SQL tests, we do mock out the sql execution engine. For pandas everything is in memory and there is no external service to mock out. I believe you can use `execution_engine_override` as a hook to use a fake execution engine.

File: tests/experimental/datasources/test_utils.py
Comment (billdirks) [reply tests/experimental/datasources/test_utils.py]: This should be our e2e test so the goal here is to test that we can go from a zep datasource and generate datadocs and they look correct (ie we see no expectations failing, metrics look correct, the datadocs look right, etc). It is a lot of stuff but I wanted something that was comprehensive since this was a new entry point to this workflow.
The goal was to have 1 e2e test for each major workflow (computing expectations, checkpoints, data assistants) and then lots of unit tests for particular functionality.

**Review by billdirks — COMMENTED** [review]
Thanks for doing this Alex! Overall it looks fine. There is some code copy/pasted which is ok but it may be nice to find a shared location for code shared between pandas and spark. The regex matching and filesystem code seems datasource agnostic.
I also left some questions about tests. I would like to make it more obvious which code should be used in integration tests only. Since you are breaking up the integration test code, I have a suggestion in `conftest.py` about organization.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py]: @cdkini This has been done.  Thanks.

File: great_expectations/experimental/datasources/interfaces.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/interfaces.py]: @cdkini This has been done.  Thanks.

File: great_expectations/experimental/datasources/dynamic_pandas.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/dynamic_pandas.py:126]: @billdirks We did indeed discuss this during our sync up (after the prioritization meeting) and are in agreement.  Thanks.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py]: @billdirks The JIRA ticket has been filed; the comment will be deleted (in next commit).  Thanks.

File: great_expectations/experimental/datasources/pandas_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/pandas_datasource.py]: @billdirks Good point -- fixing (will be in next commit).  Thanks.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py]: @billdirks This has been done.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py:35]: @billdirks I am not comfortable doing it yet; perhaps soon (when moving towards build file-path data assets as I had attempted in #6928).  Thanks.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py]: @billdirks Good point; I fixed it (in next commit).  Thanks.

File: tests/experimental/datasources/test_spark_datasource.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_spark_datasource.py]: @billdirks Let us discuss -- it seems that `file_relative_path()` is a useful general utility that accomplishes this.  Thanks.

File: tests/experimental/datasources/conftest.py
Comment (billdirks) [reply tests/experimental/datasources/conftest.py:1]: @alexsherstinsky This still needs to be moved to the integration folder.

File: tests/experimental/datasources/tasks.py
Comment (billdirks) [inline tests/experimental/datasources/tasks.py]: Can you revert this back? One thing I really don't like in our codebase is the transformation of objects into strings and then back into objects again. If we want a Path object we should avoid turning it into a string unless that is needed by a third party lib. Alternatively you coud update `file_relative_path` to return a `Path` object but that is probably better done outside of this PR.

File: tests/experimental/datasources/test_config.py
Comment (billdirks) [inline tests/experimental/datasources/test_config.py]: Similar comment here as above. We should avoid turning things into strings.

File: tests/experimental/datasources/tasks.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/tasks.py]: @billdirks No, I should not modify `file_relative_path()` -- so I will revert the change -- that is more proper.  One moment.

File: tests/experimental/datasources/test_config.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_config.py]: @billdirks Reverting (commit coming soon).  Thanks.

File: tests/experimental/datasources/conftest.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/conftest.py:1]: @billdirks This has been done and a higher level "conftest.py" created (containing the fixtures for the unit tests that need them).  Thanks.

File: tests/experimental/datasources/test_spark_datasource.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_spark_datasource.py]: @billdirks This has been reverted and standardized in an elegant, consistent way (now I can clearly see that it is indeed amenable to a utility function that would stop short of outputting a string -- for another time).  Thanks.

File: tests/experimental/datasources/tasks.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/tasks.py]: @billdirks This has been taken care of.  Thanks.

File: tests/experimental/datasources/test_config.py
Comment (alexsherstinsky) [reply tests/experimental/datasources/test_config.py]: @billdirks This has been taken care of. Thanks.

**Review by billdirks — APPROVED** [review]
Thanks @alexsherstinsky!

Please do the follow up items we discussed after this PR merges:
* Coordinate with @NathanFarmer to make sure the changes from his PR apply to this.
* Look at mocking the pyspark engine in the unit tests so we can run them without pyspark installed.
* logger -> LOGGER (this wasn't in a discussion with me but was in a slack thread)

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (Kilo59) [reply great_expectations/experimental/datasources/spark_datasource.py]: That seems like unnecessary churn that's going to lead to conflicts. IMO it would be better to drop/revert the changes to `LOGGER`.

Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/spark_datasource.py]: @Kilo59 @billdirks @cdkini While we previously had a motion to replace `logger` (all lowercase) with `LOGGER` (all uppercase) and even though doing so has not caused any checks to fail, not everybody is on-board with this change at the present time.  Hence, per the recommendation by @Kilo59 I am reverting it (we can always apply it at a later date, should we rationalize that this aspect of adhering to the Google Python style is beneficial to GX OSS.  Thank you. /cc @donaldheppner @5u2ie

**Pull request merged**
